src/hotspot/share/opto/divnode.cpp

```diff
@@ -1084,12 +1149,97 @@ const Type* ModINode::Value(PhaseGVN* phase) const {
 
 //=============================================================================
 //------------------------------Idealize---------------------------------------
-Node *UModINode::Ideal(PhaseGVN *phase, bool can_reshape) {
+
+template <typename TypeClass, typename Signed>
+static Node* unsigned_mod_ideal(PhaseGVN* phase, bool can_reshape, Node* mod) {
   // Check for dead control input
-  if( in(0) && remove_dead_region(phase, can_reshape) )  return this;
+  if (mod->in(0) && mod->remove_dead_region(phase, can_reshape)) {
+    return mod;
+  }
+  // Don't bother trying to transform a dead node
+  if (mod->in(0) && mod->in(0)->is_top()) {
+    return nullptr;
+  }
+
+  // Get the modulus
+  const Type* t = phase->type(mod->in(2));
+  if (t == Type::TOP) {
+    return nullptr;
+  }
+  const TypeClass* ti = t->is<TypeClass>();
+
+  // Check for useless control input
+  // Check for excluding mod-zero case
+  if (mod->in(0) && (ti->_hi < 0 || ti->_lo > 0)) {
+    mod->set_req(0, nullptr); // Yank control input
+    return mod;
+  }
+
+  if (!ti->is_con()) {
+    return nullptr;
+  }
+  Signed con = ti->get_con();
+
+  if (con == 1) {
+    return ConNode::make(TypeClass::ZERO);
```

Comment:
This should be in `Value` instead.

Reply:
This is analogous to code for ModI/LNode::Ideal. I'll file an RFE that this should be changed in all locations

Reply:
Fixed